test/integration/internal_listener_integration_test.cc

```diff
@@ -0,0 +1,113 @@
+#include "envoy/config/bootstrap/v3/bootstrap.pb.h"
+#include "envoy/config/core/v3/config_source.pb.h"
+#include "envoy/config/core/v3/grpc_service.pb.h"
+#include "envoy/extensions/filters/network/tcp_proxy/v3/tcp_proxy.pb.h"
+#include "envoy/network/connection.h"
+#include "envoy/service/discovery/v3/discovery.pb.h"
+
+#include "source/common/config/api_version.h"
+
+#include "test/common/grpc/grpc_client_integration.h"
+#include "test/integration/base_integration_test.h"
+#include "test/test_common/network_utility.h"
+#include "test/test_common/resources.h"
+
+#include "absl/strings/str_cat.h"
+#include "gmock/gmock.h"
+#include "gtest/gtest.h"
+
+namespace Envoy {
+namespace {
+
+class InternalListenerIntegrationTest : public testing::TestWithParam<Network::Address::IpVersion>,
+                                        public BaseIntegrationTest {
+public:
+  InternalListenerIntegrationTest()
+      : BaseIntegrationTest(GetParam(), ConfigHelper::baseConfig() + +R"EOF(
+    filter_chains:
+    - filters:
+      - name: envoy.filters.network.tcp_proxy
+        typed_config:
+          "@type": type.googleapis.com/envoy.extensions.filters.network.tcp_proxy.v3.TcpProxy
+          stat_prefix: tcp_stats
+          cluster: cluster_0
+)EOF") {}
+
+  void initialize() override {
+    config_helper_.renameListener("tcp");
+    config_helper_.addConfigModifier([](envoy::config::bootstrap::v3::Bootstrap& bootstrap) {
+      auto& listener = *bootstrap.mutable_static_resources()->mutable_listeners(0);
+      listener.mutable_address()->mutable_envoy_internal_address()->set_server_listener_name(
+          "internal_listener");
+      listener.mutable_internal_listener();
```

Comment:
Why is this needed? Shouldn't it be enough to set the address type to internal?

Reply:
That's a good point. The internal listener field was supposed to declare some settings, e.g. loop prevention check, also an indicator that Envoy would ignore the tcp socket options. I tend to agree with you that this internal options should be optional

What do you think the major drawback here?

Reply:
If the user needs to specify a field in the address (setting the type to internal), it should not be required to set this message; that would mean the user has to essentially set the same logical setting in two places. If there are config settings needed for the internal listener, I think it makes sense to put them here. But if there aren't, and the address is set to internal, that should work.

Reply:
I made the change in https://github.com/envoyproxy/envoy/pull/18104/commits/3f6407b5bdd5eed9d9ab94653668329da701b20d

Basically, this protobuf internal listener config is not required. The listen address. BTW:  I also add the test case that this field is present, the listen address must be internal type.